src/librustc_mir/build/expr/as_temp.rs

```diff
@@ -85,8 +85,10 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
 
         unpack!(block = this.into(&Place::Local(temp), block, expr));
 
-        // In constants, temp_lifetime is None. We do not drop anything because
-        // values with a destructor will simply be leaked in constants.
+        // In constants, temp_lifetime is None for temporaries that live for the
+        // entire constant. Thus we do not drop these temporaries and simply leak them.
```

Comment:
Not "entire constant". The semantics of shallow temporaries in the RHS of a binding is that they escape into the *ambient* scope. For a `const`/`static`, this ambient scope is `'static`.